### ISSUE TITLE
`multiple definition` issue when including hpp header files

### ISSUE THREAD

**Issue by tree786isback** [Immediate-Mode-UI/Nuklear#425]
Hello

I am writing a GUI rendering class which uses nuklear as the GUI backend. My header file contains various nuklear structs (`struct nk_context *ctx` etc), and it has a setup like this:
```
#ifndef NUKLEAR_GUI_HPP 
#define NUKLEAR_GUI_HPP //header include guard

#define NK_INCLUDE_FIXED_TYPES
#define NK_INCLUDE_STANDARD_IO
#define NK_INCLUDE_STANDARD_VARARGS
#define NK_INCLUDE_DEFAULT_ALLOCATOR
#define NK_INCLUDE_VERTEX_BUFFER_OUTPUT
#define NK_INCLUDE_FONT_BAKING
#define NK_INCLUDE_DEFAULT_FONT
#define NK_IMPLEMENTATION
#define NK_GLFW_GL3_IMPLEMENTATION
#define NK_KEYSTATE_BASED_INPUT
#define INCLUDE_STYLE
#define WINDOW_WIDTH 1200
#define WINDOW_HEIGHT 800

#define MAX_VERTEX_BUFFER 512 * 1024
#define MAX_ELEMENT_BUFFER 128 * 1024

#include <GL/glew.h>
#include <GLFW/glfw3.h>

#include "nuklear.h"
#include "nuklear_glfw_gl3.h"
#include "style.c"

//NUKLEAR CLASS
class NuklearGui{
public:
  struct nk_glfw glfw = {0};
  int* width;
  int* height;
  struct nk_context *ctx;
  struct nk_colorf bg;
  struct nk_command_buffer *canvas;
  NuklearGui();
}


#endif // NUKLEAR_GUI_HPP
```
However, upon compiling, I get a huge backlog of `multiple definition` errors, like this:
```
multiple definition of `nk_glfw3_font_stash_end(nk_glfw*)'; CMakeFiles/binary.dir/src/Application.cpp.o:Application.cpp:(.text+0x4b5e0):
```
I don't understand why this is happening, as I've already included my own header guards. Any ideas what i could be doing wrong?

**Comment by tree786isback** [comment]
Well I've managed to overcome this by defining the nuklear structs as pointers. But now I'm getting all kinds of segfaults once it compiles and runs

**Comment by dumblob** [comment]
These errors usually stem from the fact that `nuklear.h` is actually being included more than once. Could you investigate whether there are any duplicated symbols in the output of `gcc -E ...`?

**Comment by dumblob** [comment]
> Ah I see, that's correct. I thought that indirect includes didn't count. How can I overcome this issue then? As I need this `Application.cpp` in my other scripts, but I'd also like to add GUI rendering inside the `Application` class as well.

It's simpler than you'd think :wink:. Just follow the readme - i.e. define nuklear implementation **only once**. That can be guaranteed only if you place this define into the very root of the whole include hierarchy (probably into your `MyApplication.c`) before any of your includes. And it's done :wink:.

Of course you must **not** define it anywhere else!

**Comment by tree786isback** [comment]
Strangely enough, I have now included the define and nuklear header in only one cpp file, but when I define a nuklear struct (`struct nk_context *ctx`) in another hpp file, I get this error:
```
/usr/bin/ld: CMakeFiles/binary.dir/src/MyApplication.cpp.o:(.bss+0x0): multiple definition of `ctx'; CMakeFiles/binary.dir/src/Application.cpp.o:(.bss+0x4c88): first defined here
/usr/bin/ld: CMakeFiles/binary.dir/src/main.cpp.o:(.bss+0x0): multiple definition of `ctx'; CMakeFiles/binary.dir/src/Application.cpp.o:(.bss+0x4c88): first defined here
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/binary.dir/build.make:170: binary] Error 1
make[1]: *** [CMakeFiles/Makefile2:256: CMakeFiles/binary.dir/all] Error 2
make: *** [Makefile:136: all] Error 2

```
I don't understand why it's saying I've included context multiple times

**Comment by dumblob** [comment]
Hm, interesting. Could you provide the whole source code?

**Comment by tree786isback** [comment]
Sure thing, here is the full project:
Excuse the messy setup, the files of relevance are `Application.cpp` and `Application.hpp` located in the src folder. I define `  struct nk_context *ctx;` in the `Application.hpp` file, and the nuklear headers and defines are declared inside the `Application.cpp` file
[nuklear_gui_app.zip](https://github.com/Immediate-Mode-UI/Nuklear/files/8248100/nuklear_gui_app.zip)

**Comment by tree786isback** [comment]
Oh I didn't notice that, thanks for spotting. I cleaned that up compiled, but it still gives me the error though
However I have found that adding `struct nk_context *ctx` directly inside my `NuklearGui` class fixes the compile issue. Like this:
```
class NuklearGui{
public:
  struct nk_context *ctx;
  struct nk_glfw* glfw;
  int* width;
  int* height;
  // struct nk_colorf* bg;
  struct nk_command_buffer *canvas;

  NuklearGui(GLFWwindow* window);
  void render();
};
```
But now another issue has come up -- I start getting segfaults when I init the `NuklearGui` class as a ptr, then call on the ctx variable inside

**Comment by tree786isback** [comment]
Here is my current setup. In my `Application.hpp`, I have the declare the `NuklearGui` class like this:
```
class NuklearGui{
public:
  struct nk_context *ctx;
  struct nk_glfw* glfw;
  int* width;
  int* height;
  struct nk_colorf* bg;
  struct nk_command_buffer *canvas;

  void init(GLFWwindow* window);
  void render();
};
```
And in my `Application.cpp`, I define it's init function like this:
```
struct nk_glfw glfw_init = {0};

NuklearGui nuk_gui;

void NuklearGui::init(GLFWwindow* window){
  glfw = &glfw_init;

  ctx = nk_glfw3_init(glfw, window, NK_GLFW3_INSTALL_CALLBACKS);

  struct nk_font_atlas *atlas;
  nk_glfw3_font_stash_begin(glfw, &atlas);
  nk_glfw3_font_stash_end(glfw);
  bg->r = 0.10f, bg->g = 0.18f, bg->b = 0.24f, bg->a = 0.0f;
}

```
However, I get a segfault when setting the `bg` variable:
```
bg->r = 0.10f, bg->g = 0.18f, bg->b = 0.24f, bg->a = 0.0f;
```
I have ran `gdb` to find more info on this segfault, and all it says it this
```
Program received signal SIGSEGV, Segmentation fault.
0x00005555555f8bb9 in NuklearGui::init(GLFWwindow*) ()

```
Not very useful, but I don't get why deref'ing the `struct nk_colorf* bg` and assigning it the floats causes segfaults. Any ideas on this? I have mostly dealt with c++, so not sure if this is a misunderstading on how to use C functions on my end

**Comment by dumblob** [comment]
Ok, I've taken a look at the source and the original issue was `struct nk_context *ctx;` being defined in a `.hpp` file which was included in more than one `*.cpp` files which were compiled separately and thus both object files contained this symbol (and the space for the variable) and the linker didn't know which one to use. Simply commenting that line solved the compilation issue (because `ctx` was not being used anywhere - all was commented).

Regarding dereferencing of `struct nk_colorf* bg;` in your class I don't see any place where you allocate memory to the pointer - so therefore it segfaults if you try to dereference it. So I'd turn it into `struct nk_colorf bg;` (i.e. struct instead of a pointer to struct).

**Comment by tree786isback** [comment]
Thanks a lot for the explanation and help!
> So I'd turn it into struct nk_colorf bg; (i.e. struct instead of a pointer to struct).

I was getting `incomplete type` errors when I was initially  declaring as a full type, like this:
```
/home/mm/Desktop/nuklear_gui_app/src/Application.hpp:38:20: error: field ‘bg’ has incomplete type ‘nk_colorf’
   38 |   struct nk_colorf bg;
      |                    ^~
/home/mm/Desktop/nuklear_gui_app/src/Application.hpp:38:10: note: forward declaration of ‘struct nk_colorf’
   38 |   struct nk_colorf bg;
      |          ^~~~~~~~~
```
So I instead converted it to a pointer to allow it to compile. 
>in your class I don't see any place where you allocate memory to the pointer

I am allocating it mem like this (if I understood correctly)
```
class NuklearGui{
public:
  struct nk_context *ctx;
  struct nk_glfw* glfw;
  int* width;
  int* height;
  struct nk_colorf* bg; //declare ptr to bg var
  struct nk_command_buffer *canvas;
  void init(GLFWwindow* window);
  virtual void render();
};

```

**Comment by dumblob** [comment]
You're compiling `Application.o` separately and for that you need to have the symbols somewhere in each file which uses them. So you need to copy (and maintain them **identical**!) the whole block of defines and the `nuklear.h` include from `Application.cpp` to `Application.hpp`:

```c
#define NK_INCLUDE_FIXED_TYPES
#define NK_INCLUDE_STANDARD_IO
#define NK_INCLUDE_STANDARD_VARARGS
#define NK_INCLUDE_DEFAULT_ALLOCATOR
#define NK_INCLUDE_VERTEX_BUFFER_OUTPUT
#define NK_INCLUDE_FONT_BAKING
#define NK_INCLUDE_DEFAULT_FONT
//#define NK_IMPLEMENTATION
//#define NK_GLFW_GL3_IMPLEMENTATION
#define NK_KEYSTATE_BASED_INPUT
#define INCLUDE_STYLE

#include "nuklear.h"
//#include "nuklear_glfw_gl3.h"
//#include "style.c"
```

**Comment by tree786isback** [comment]
Super, many thanks again for the dedicated help! It works very well now!
